test/tls13groupselection_test.c

```diff
@@ -0,0 +1,292 @@
+    if (ssl_or_ctx == WORK_ON_SSL_OBJECT) {
+        if (current_test_vector->client_groups != NULL)
+            TEST_true_or_end(SSL_set1_groups_list(clientssl, current_test_vector->client_groups));
+        if (current_test_vector->server_groups != NULL)
+            TEST_true_or_end(SSL_set1_groups_list(serverssl, current_test_vector->server_groups));
+        TEST_true_or_end(SSL_set_min_proto_version(clientssl, TLS1_3_VERSION));
+        TEST_true_or_end(SSL_set_min_proto_version(serverssl, TLS1_3_VERSION));
+        if (current_test_vector->preference == SERVER_PREFERENCE)
+            SSL_set_options(serverssl, SSL_OP_CIPHER_SERVER_PREFERENCE);
+    }
+
+    /* Creating a test connection */
+    TEST_true_or_end(create_ssl_connection(serverssl, clientssl, SSL_ERROR_NONE));
+
+    /*
+     * Checking that the negotiated group matches our expectation
+     * and must be identical on server and client
+     */
+    negotiated_group_client = SSL_get_negotiated_group(clientssl);
+    negotiated_group_server = SSL_get_negotiated_group(serverssl);
+    if (!TEST_int_eq(negotiated_group_client, negotiated_group_server))
+        goto end;
+    if (TEST_int_eq(negotiated_group_client, OBJ_sn2nid(current_test_vector->expected_group)))
+        ok = 1;
+
+end:
+    SSL_free(serverssl);
+    SSL_free(clientssl);
+    SSL_CTX_free(server_ctx);
+    SSL_CTX_free(client_ctx);
+    return ok;
+}
+#endif /* !defined(OPENSSL_NO_EC) */
+
+static int tls13groupselection_test(void)
+{
+    int testresult = 1; /* Assume the test will succeed */
+#if !defined(OPENSSL_NO_EC) && !defined(OPENSSL_NO_TLS1_3)
+    unsigned long i;
+
+    /*
+     * Call the code under test, once such that the ssl object is used,
+     * once such that the ctx is usedIf any of the tests fail (= return 0),
+     * the end result will be 0 thanks to multiplication
+     */
+    for (i = 0;
+         i < sizeof(tls13groupselection_tests) / sizeof(tls13groupselection_tests[0]);
+         i++) {
+        int res = 0;
+
+        fprintf(stdout, "==> Running TLSv1.3 test %lu\n", i);
+        res = run_13groupselection(&tls13groupselection_tests[i],
+                                   WORK_ON_SSL_OBJECT);
+        if (!res)
+            fprintf(stdout, "====> [ERROR] TLSv1.3 test %lu with WORK_ON_SSL_OBJECT failed\n", i);
+        testresult *= res;
+        res = run_13groupselection(&tls13groupselection_tests[i],
+                                   WORK_ON_CONTEXT);
+        if (!res)
+            fprintf(stdout, "====> [ERROR] TLSv1.3 test %lu with WORK_ON_CONTEXT failed\n", i);
+        testresult *= res;
+    }
+
+#endif /* !defined(OPENSSL_NO_EC) */
+
+    return testresult;
+}
+
+int setup_tests(void)
+{
+
+#if !defined(OPENSSL_NO_EC) && !defined(OPENSSL_NO_TLS1_3)
+    if (!TEST_ptr(cert = test_get_argument(0))
+            || !TEST_ptr(privkey = test_get_argument(1)))
+        return 0;
+#endif /* !defined(OPENSSL_NO_EC) */
+
+    ADD_TEST(tls13groupselection_test);
+    return 1;
```

Comment:
This basically seems virtually identical to the TLSv1.2 test. I'm not sure why we need two copies of essentially the same thing. I'd also consider putting this in sslapitest. Not sure why its been broken out.

Reply:
@mattcaswell thanks for all the great comments. tls12groupselection_test was something I quickly threw in as I wanted to try just to ensure that when it was tls 1.2, the code changes didn't affect processing the group list. I'm happy to remove tls12groupselection_test as there are already similar tests to cover this scenario (apologies as I am just getting familiar with some of the codebase). The tls13groupselection_test will get invalid tests on the next drop of the code as that is on the list of things to be done in line with handling the current set of comments and changing the delimiter